##**Function to convert a Microsoft Excel Epoch date to a "YYYY-MM-DD" date Format.**

##**Books borrowed and returned were in 2023.**

In [3]:
import csv
import datetime

def excel_to_date(excel_date):
    """
    Convert a date represented as the number of days since the Excel epoch to a string in the format 'YYYY-MM-DD'.

    Parameters:
        excel_date (int): The number of days since the Excel epoch (December 30, 1899) to be converted.

    Returns:
        str: A string in the format 'YYYY-MM-DD' representing the date corresponding to the input number of days.
    """
    excel_epoch_start = datetime.datetime(1899, 12, 30)
    py_date = excel_epoch_start + datetime.timedelta(days=excel_date)
    return py_date.strftime('%Y-%m-%d')

# Read book loans CSV file
book_loans = []
with open('bookloans (1).csv', newline='') as csvfile:
    reader = csv.reader(csvfile)
    for row in reader:
        book_loans.append(row)

# Remove header row
header = book_loans.pop(0)

# Convert dates
for loan in book_loans:
    loan[2] = excel_to_date(int(loan[2]))
    loan[3] = excel_to_date(int(loan[3]))

# Filter loans for 2023
book_loans_2023 = [loan for loan in book_loans if '2023' in loan[2] or '2023' in loan[3]]

# Add header
header = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Print formatted header
print("{:<15} {:<15} {:<15} {:<15}".format(*header))

# Print formatted loans
for loan in book_loans_2023:
    print("{:<15} {:<15} {:<15} {:<15}".format(*loan))


book_number     member_number   date_of_loan    date_of_return 
31              192             2023-01-01      2023-01-19     
57              199             2023-01-01      2023-01-21     
100             140             2023-01-01      2023-01-15     
100             39              2023-01-01      2023-01-08     
114             196             2023-01-01      2023-01-12     
114             171             2023-01-01      2023-01-20     
138             109             2023-01-01      2023-01-04     
10              150             2023-01-02      2023-01-15     
12              192             2023-01-02      2023-01-13     
28              165             2023-01-02      2023-01-16     
41              30              2023-01-02      2023-01-11     
105             112             2023-01-02      2023-01-04     
140             171             2023-01-02      2023-01-13     
75              93              2023-01-03      2023-01-21     
94              186             2023-01-

# **Task 1**

## **PART I**

### **books bought multiple times**

In [14]:
import csv

# Read book loans CSV file and add header
book_loans = []
with open('/content/bookloans (1).csv', newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=["book_number", "member_number", "date_of_loan", "date_of_return"])
    next(reader)  # Skip header row
    for row in reader:
        book_loans.append(row)

# Count occurrences of each book
book_counts = {}
for loan in book_loans:
    book_number = loan["book_number"]
    if book_number in book_counts:
        book_counts[book_number] += 1
    else:
        book_counts[book_number] = 1

# Filter books that have been purchased more than once
multi_purchase_books = {book: count for book, count in book_counts.items() if count > 1}

# Sort book numbers in ascending order
sorted_books = sorted(multi_purchase_books.items(), key=lambda x: int(x[0]))

# Display the books that have been purchased multiple times
print("        book_number    purchase_count")
for index, (book, count) in enumerate(sorted_books):
    print(f"{index}\t\t{book}\t\t{count}")


        book_number    purchase_count
0		1		22
1		2		21
2		3		8
3		4		4
4		5		22
5		6		18
6		7		22
7		8		15
8		9		9
9		10		16
10		11		20
11		12		22
12		14		18
13		15		20
14		16		17
15		17		18
16		18		4
17		19		13
18		20		24
19		21		21
20		22		21
21		23		14
22		24		20
23		25		11
24		26		11
25		27		18
26		28		8
27		29		7
28		30		20
29		31		22
30		32		21
31		33		15
32		34		10
33		35		2
34		36		23
35		37		24
36		38		28
37		39		24
38		40		3
39		41		20
40		42		5
41		43		25
42		44		5
43		45		11
44		46		7
45		47		24
46		48		25
47		49		20
48		50		5
49		51		25
50		52		7
51		53		20
52		54		23
53		55		16
54		56		24
55		57		2
56		58		23
57		59		2
58		60		21
59		61		27
60		62		21
61		63		23
62		64		11
63		65		19
64		66		25
65		67		2
66		68		8
67		69		7
68		70		7
69		71		3
70		72		27
71		73		6
72		74		4
73		75		17
74		76		17
75		77		5
76		78		24
77		79		23
78		80		15
79		81		12
80		82		12
81		83		24
82		84		17
83		85		28
84		86		4
85		87		23
86		88		23
87		89		23
88		90		16
89		91		20
90		92		24
91		

**Top 5 most popular books**

In [17]:
import csv

# Read book loans CSV file and add header
book_loans = []
with open('/content/bookloans (1).csv', newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=["book_number", "member_number", "date_of_loan", "date_of_return"])
    next(reader)  # Skip header row
    for row in reader:
        book_loans.append(row)

# Count occurrences of each book
book_counts = {}
for loan in book_loans:
    book_number = loan["book_number"]
    if book_number in book_counts:
        book_counts[book_number] += 1
    else:
        book_counts[book_number] = 1

# Sort books by purchase count in descending order
sorted_books = sorted(book_counts.items(), key=lambda x: x[1], reverse=True)

# Display the top 5 most sold books
print("Top 5 most sold books:")
print("serial no.\tbook_number\tpurchase_count")
for index, (book, count) in enumerate(sorted_books[:5], start=1):
    print(f"{index}\t\t{book}\t\t{count}")


Top 5 most sold books:
serial no.	book_number	purchase_count
1		85		28
2		38		28
3		119		28
4		61		27
5		72		27


### **The 5 most popular books are:**
```
Book Number       Book Title                        Total bought
1. 85       |     The Great Indian Novel           |     28
2. 119      |     Karl Marx Biography              |     28
3. 38       |     False Impressions                |     28
4. 72       |     A Prisoner Of Birth              |     27
5. 61       |     A Modern Approach Computer Vision|     27
```



**least popular books**

In [19]:
import csv

# Read book loans CSV file and add header
book_loans = []
with open('/content/bookloans (1).csv', newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=["book_number", "member_number", "date_of_loan", "date_of_return"])
    next(reader)  # Skip header row
    for row in reader:
        book_loans.append(row)

# Count occurrences of each book
book_counts = {}
for loan in book_loans:
    book_number = loan["book_number"]
    if book_number in book_counts:
        book_counts[book_number] += 1
    else:
        book_counts[book_number] = 1

# Sort books by purchase count in descending order
sorted_books = sorted(book_counts.items(), key=lambda x: x[1], reverse=False)

# Display the top 5 most sold books
print("Top 5 most sold books:")
print("serial no.\tbook_number\tpurchase_count")
for index, (book, count) in enumerate(sorted_books[:5], start=1):
    print(f"{index}\t\t{book}\t\t{count}")


Top 5 most sold books:
serial no.	book_number	purchase_count
1		13		1
2		123		1
3		57		2
4		125		2
5		35		2


### **The 5 least popular books are:**

```
Book Number      Book Title                         Total bought
1. 13      |     Birth Of A Theorem                |     1
2. 123     |     Selected Short Stories            |     1
3. 35      |     Surely You'Re Joking Mr Feynman   |     2
4. 57      |     Textbook Of Economic Theory       |     2
5. 59      |     Learning Opencv                   |     2
```



## **PART II**

# **book appears on the loan list in 2023.**   
**reverse order of their popularity.**

**all rows dislayed**

In [33]:
import csv
import datetime

# Function to convert Excel date to 'YYYY-MM-DD' format
def excel_to_date(excel_date: int) -> str:
    excel_epoch_start = datetime.datetime(1899, 12, 30)
    py_date = excel_epoch_start + datetime.timedelta(days=excel_date)
    return py_date.strftime('%Y-%m-%d')

# Open the CSV file and read data
with open("/content/bookloans (1).csv", mode='r') as file:
    reader = csv.reader(file)
    header = next(reader)  # Skip header row if there is one, otherwise remove this line

    # Define a dictionary to count book borrowings
    book_count = {}

    for row in reader:
        book_number = row[0]
        date_of_loan = excel_to_date(int(row[2]))

        # Filter data for the year 2023
        if "2023-01-01" <= date_of_loan <= "2023-12-31":
            if book_number in book_count:
                book_count[book_number] += 1
            else:
                book_count[book_number] = 1

# Sort book numbers by borrow count in descending order
sorted_books = sorted(book_count.items(), key=lambda item: item[1], reverse=True)

# Print results
print("Number of days for which each book was borrowed in 2023:")
print("               book_number           maximum duration of each Book rented")
for index,     (book, count) in enumerate(sorted_books):
    print(f"{index:5}            {book:12}     {count:35}")



Number of days for which each book was borrowed in 2023:
               book_number           maximum duration of each Book rented
    0            85                                                28
    1            38                                                28
    2            119                                               28
    3            61                                                27
    4            72                                                27
    5            105                                               25
    6            94                                                25
    7            48                                                25
    8            66                                                25
    9            132                                               25
   10            43                                                25
   11            51                                                25
   12            114         

In 2023, the library counted how many times each book was borrowed to see which ones were popular among readers. This helps the library decide which books to keep based on how often they are checked out, showing which ones people enjoy and find valuable.

# **Task 2**
###**Types of Genres and movies related to each Genre**

In [58]:
import csv
import datetime

# Define header names for the book loans data
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Load the datasets
updated_book_loans_path = '/content/bookloans (1).csv'
updated_books_path = '/content/books (1).csv'

# Read the book loans data
book_loans = []
with open(updated_book_loans_path, newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=header_names)
    next(reader)  # Skip header row
    for row in reader:
        book_loans.append(row)

# Read the books data
books = []
with open(updated_books_path, newline='') as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        books.append(row)

# Define the function to convert Excel serial date to year
def serial_date_to_string(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=int(s))
    return (excel_epoch + delta).year

# Apply conversion to date_of_loan column to filter for 2023
loans_2023 = [loan for loan in book_loans if serial_date_to_string(loan['date_of_loan']) == 2023]

# Join the filtered loan data with the books data
loans_with_genre = []
for loan in loans_2023:
    for book in books:
        if loan['book_number'] == book['Number']:
            loan['Genre'] = book['Genre']
            loans_with_genre.append(loan)
            break

# Count occurrences of each genre
genre_counts = {}
for loan in loans_with_genre:
    genre = loan['Genre']
    if genre in genre_counts:
        genre_counts[genre] += 1
    else:
        genre_counts[genre] = 1

# Sort by count in descending order
sorted_genre_counts = sorted(genre_counts.items(), key=lambda x: x[1], reverse=True)




# Display the result in a visually appealing format
print("Popularity Report of All Genres of Books in 2023")
print("-" * 45)
print("Genre\t\t\tNumber of Books")
print("-" * 45)
for genre, count in sorted_genre_counts:
    print(f"{genre.ljust(20)}\t\t{count}")


Popularity Report of All Genres of Books in 2023
---------------------------------------------
Genre			Number of Books
---------------------------------------------
Fiction             		716
Non-fiction         		591
Tech                		396
Science             		153
Philosophy          		101


The  cell displays the categories in the library long with the total numbers of books within each category.

# **Task 3**
### **Statistics on loans and late loans are lost.**

 *• Number of loans.*

*• Number of days borrowed.*

*• Number of loans returned late.*

*• Number days late.*

In [60]:
import csv
import datetime

# Define header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
book_loans = []
with open('/content/bookloans (1).csv', newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=header_names)
    next(reader)  # Skip header row
    for row in reader:
        book_loans.append(row)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=int(s))
    return excel_epoch + delta

# Convert loan and return dates to datetime objects
for loan in book_loans:
    loan['date_of_loan'] = serial_date_to_datetime(loan['date_of_loan'])
    loan['date_of_return'] = serial_date_to_datetime(loan['date_of_return'])

# Calculate total number of loans
total_loans = len(book_loans)

# Calculate total number of days borrowed
total_days_borrowed = sum((loan['date_of_return'] - loan['date_of_loan']).days for loan in book_loans if loan['date_of_return'] >= loan['date_of_loan'])

# Calculate number of loans returned late
total_books_returned_late = sum((loan['date_of_return'] - loan['date_of_loan']).days > 14 for loan in book_loans)

# Calculate total days late
total_days_late = sum(max((loan['date_of_return'] - loan['date_of_loan']).days - 14, 0) for loan in book_loans)

# Output the results
print("\033[1mTotal number of loans:\033[0m" +"          |  "+ str(total_loans))
print("\033[1mTotal number of days borrowed:\033[0m"+"  |  "+ str(total_days_borrowed))
print("\033[1mTotal loans returned late:\033[0m" +"      |  "+ str(total_books_returned_late))
print("\033[1mTotal days late:\033[0m" +"                |  "+ str(total_days_late))


Total number of loans:          |  2208
Total number of days borrowed:  |  22118
Total loans returned late:      |  612
Total days late:                |  2141


The calculations show that in total, there were 2,209 loans made from the library, Gather to 22,135 days borrowed. Among these, 613 loans were returned late, resulting in a total of 2,144 days overdue.

## **a) The average number of days that a book was borrowed.**

### ***calculating the average number of days a book was borrowed across all loans, rather than for each individual book.***

In [69]:
import csv
import datetime

# Function to convert Excel serial date to Python datetime object
def serial_date_to_datetime(serial_date):
    excel_epoch = datetime.datetime(1899, 12, 30)
    return excel_epoch + datetime.timedelta(days=serial_date)

# Path to the CSV file
csv_path = '/content/bookloans (1).csv'

# Initialize variables for calculations
total_days_borrowed = 0
total_loans = 0

# Read CSV file
with open(csv_path, mode='r') as file:
    reader = csv.reader(file)
    next(reader)  # Skip the header row

    for row in reader:
        # Convert the date columns
        date_of_loan = serial_date_to_datetime(int(row[2]))
        date_of_return = serial_date_to_datetime(int(row[3]))

        # Calculate days borrowed, ensure no negative days
        days_borrowed = (date_of_return - date_of_loan).days
        days_borrowed = max(days_borrowed, 0)  # No negative days

        # Update totals
        total_days_borrowed += days_borrowed
        total_loans += 1

# Calculate the average days borrowed
average_days_borrowed = total_days_borrowed / total_loans if total_loans > 0 else 0

# Output the results
print("\033[1mAverage days a book was borrowed:\033[0m", average_days_borrowed)


Average days a book was borrowed: 10.017210144927537


it shows How many days a each book have been minimum borrowed and it shows the average of duration of all books in library.

## **b) The percentage proportion of books returned late.**

### ***Count of all loan transactions. Computes the days each book was borrowed. Determines if the loan was returned late by checking if days_borrowed is greater than 14. Calculates the proportion of late returns relative to the total number of loans and formats it as a percentage.***

In [70]:
import csv
import datetime

# Define header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
updated_book_loans = []
with open('/content/bookloans (1).csv', newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=header_names)
    next(reader)  # Skip header row
    for row in reader:
        updated_book_loans.append(row)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=int(s))
    return excel_epoch + delta

# Apply conversion to the loan and return dates
for loan in updated_book_loans:
    loan['date_of_loan'] = serial_date_to_datetime(loan['date_of_loan'])
    loan['date_of_return'] = serial_date_to_datetime(loan['date_of_return'])

# Calculate total number of loans
total_loans = len(updated_book_loans)

# Calculate total number of days borrowed, ensure no negative days are counted
for loan in updated_book_loans:
    loan['days_borrowed'] = max((loan['date_of_return'] - loan['date_of_loan']).days, 0)

# Calculate number of loans returned late
total_books_returned_late = sum(loan['days_borrowed'] > 14 for loan in updated_book_loans)

# Calculate the percentage proportion of books returned late
percentage_late = (total_books_returned_late / total_loans * 100) if total_loans > 0 else 0

# Output the result
print("\033[1mPercentage of books returned late:\033[0m", f"{percentage_late:.2f}%")


Percentage of books returned late: 27.72%


The result shows in average percentage that from library how many books are returned late. The 27.75% books from total books are returned late in library.

## **c) The average late period of books returned late.**

### ***calculate the average late period of books returned late. determine how many days late each book was, and then calculate the average of these days.***

In [71]:
import csv
import datetime

# Define header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
updated_book_loans = []
with open('/content/bookloans (1).csv', newline='') as csvfile:
    reader = csv.DictReader(csvfile, fieldnames=header_names)
    next(reader)  # Skip header row
    for row in reader:
        updated_book_loans.append(row)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=int(s))
    return excel_epoch + delta

# Apply conversion to the loan and return dates
for loan in updated_book_loans:
    loan['date_of_loan'] = serial_date_to_datetime(loan['date_of_loan'])
    loan['date_of_return'] = serial_date_to_datetime(loan['date_of_return'])

# Calculate total number of loans returned late
total_books_returned_late = 0
total_days_late = 0

for loan in updated_book_loans:
    days_borrowed = (loan['date_of_return'] - loan['date_of_loan']).days
    days_late = max(days_borrowed - 14, 0)
    if days_late > 0:
        total_books_returned_late += 1
        total_days_late += days_late

# Calculate the average late period of books returned late
average_late_days = (total_days_late / total_books_returned_late) if total_books_returned_late > 0 else 0

# Output the results
print(f" \033[1mbooks returned late:\033[0m  {total_books_returned_late}")  # Number of loans returned late
print(f" \033[1mAverage:\033[0m              {average_late_days:.1f}")  # Average late period of books returned late


 books returned late:  612
 Average:              3.5


Out of the books that were brought back late, they were typically delayed by about 3.5 days on average. So, if a book was returned late, it was usually about 3.5 days overdue. And total late days of overall book is 613.  

# **Comment on data**

Markdown is beneficial for organizing and documenting code, providing clarity through formatting. It helps me structure explanations, outline code sections, and highlight important details for better readability. However, the data files provided lack in headers and proper database format, making it challenging to extract information efficiently. This increases the complexity of data processing and querying tasks. Despite these limitations. I implement functions for genre popularity, loan statistics, and late returns. I comment on the code for clarit. Markdown helps in presenting my approach and insights effectively, facilitating understanding for both myself and assessors.